apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs

```diff
@@ -0,0 +1,393 @@
+//! Code for handling [`ProtocolTxType::EthereumEvents`] transactions.
+mod events;
+mod read;
+mod update;
+mod utils;
+
+use std::collections::{BTreeSet, HashMap, HashSet};
+
+use borsh::{BorshDeserialize, BorshSchema, BorshSerialize};
+use eyre::{eyre, Result};
+use namada::ledger::eth_bridge::storage::eth_msgs::{self, Keys};
+use namada::ledger::pos::types::{VotingPower, WeightedValidator};
+use namada::types::address::Address;
+use namada::types::ethereum_events::EthereumEvent;
+use namada::types::storage::{self, Epoch};
+use namada::types::transaction::TxResult;
+use namada::types::vote_extensions::ethereum_events::MultiSignedEthEvent;
+use num_rational::Ratio;
+
+use super::store::{Store, StoreExt};
+
+/// Represents an event stored under `eth_msgs`
+#[derive(
+    Clone, Debug, PartialEq, Eq, BorshSerialize, BorshDeserialize, BorshSchema,
+)]
+pub struct EthMsg {
+    /// The event being stored
+    pub body: EthereumEvent,
+    /// The total voting power that's voted for this event across all epochs
+    pub voting_power: (u64, u64),
+    /// The addresses of validators that voted for this event, in sorted order.
+    pub seen_by: Vec<Address>,
+    /// Whether this event has been acted on or not
+    pub seen: bool,
+}
+
+/// Represents an Ethereum event being seen by some validators
+#[derive(
+    Debug,
+    Clone,
+    Ord,
+    PartialOrd,
+    PartialEq,
+    Eq,
+    Hash,
+    BorshSerialize,
+    BorshDeserialize,
+)]
+pub struct EthMsgUpdate {
+    /// the event being seen
+    pub body: EthereumEvent,
+    /// addresses of the validators who have just seen this event
+    /// we use [`BTreeSet`] even though ordering is not important here, so that
+    /// we can derive [`Hash`] for [`EthMsgUpdate`]. This also conveniently
+    /// orders addresses in the order in which they should be stored in
+    /// blockchain storage.
+    pub seen_by: BTreeSet<Address>,
+}
+
+impl From<MultiSignedEthEvent> for EthMsgUpdate {
+    fn from(
+        MultiSignedEthEvent { event, signers }: MultiSignedEthEvent,
+    ) -> Self {
+        Self {
+            body: event,
+            seen_by: signers.into_iter().collect(),
+        }
+    }
+}
+
+pub(crate) fn apply_derived_tx(
+    store: &mut impl StoreExt,
+    events: Vec<MultiSignedEthEvent>,
+) -> Result<TxResult> {
+    if events.is_empty() {
+        return Ok(TxResult::default());
+    }
+    tracing::info!(
+        ethereum_events = events.len(),
+        "Applying state updates derived from Ethereum events found in \
+         protocol transaction"
+    );
+
+    let last_epoch = store.get_last_epoch();
+    tracing::debug!(?last_epoch, "Got epoch of last block");
+
+    let active_validators = store.get_active_validators(Some(last_epoch));
```

Comment:
you should be getting these based on some block height param, especially with the new shimming stuff

Reply:
to be honest, I'm not entirely sure on this

Reply:
oh yeah good catch, this needs to change now. Leaving this comment thread open and will come back to it as need to think about it

Reply:
So in the previous design with vote extensions, using whoever the active validators were for the most recent committed block made sense, as the Ethereum events would have been voted for during the consensus round for that block. e.g. in the consensus round for proposing block `n`, validators for block `n` vote for Ethereum events. Then the Ethereum events protocol tx is made and applied in block `n+1`. The validator sets may be different between blocks `n` and `n+1` if we crossed into a new epoch, so we get the validators for block `n`. (I will try to make the code here clearer as well as all this is happening a bit implicitly)

I'm not so familiar yet with this part of the new design but that can be further issues/PRs to migrate this old code to using a specified block height (as provided in `ProtocolTxType::EthereumEvents`, I guess?). 

NB: I've now updated the PR description to make explicit this is for the old ABCI++ design.

Reply:
okay sounds good. maybe add a `TODO` here, though? so that we know we need to change this later when `abciplus` shims merge